LibraBFT/Impl/Storage/DiemDB/LedgerStore/LedgerStore.agda

```diff
@@ -3,11 +3,95 @@
+open import LibraBFT.ImplShared.Consensus.Types hiding (getEpoch)
 open import LibraBFT.Prelude
+open import Optics.All
+------------------------------------------------------------------------------
+open import Data.String                         using (String)
 
 module LibraBFT.Impl.Storage.DiemDB.LedgerStore.LedgerStore where
 
+record EpochEndingLedgerInfoIter : Set where
+  constructor EpochEndingLedgerInfoIter∙new
+  field
+    _eeliiObmLIWS : List LedgerInfoWithSignatures
+open EpochEndingLedgerInfoIter public
+
 new : LedgerStore
 new = mkLedgerStore Map.empty Map.empty nothing
+
+-- OBM-LBFT-DIFF : this entire impl file is VERY different
+
+getEpoch : LedgerStore → Version → Either ErrLog Epoch
+getEpoch self version =
+  maybeS (Map.lookup version (self ^∙ lsObmVersionToEpoch))
+         -- COMMENT FROM RUST CODE:
+         -- There should be a genesis LedgerInfo at version 0.
+         -- This normally doesn't happen.
+         -- This part of impl does not need to rely on this assumption.
+         (pure 0)
+         pure
+
+getEpochEndingLedgerInfo : LedgerStore → Version → Either ErrLog LedgerInfoWithSignatures
+getEpochEndingLedgerInfo self version = do
+  epoch ← getEpoch self version
+  case Map.lookup epoch (self ^∙ lsObmEpochToLIWS) of λ where
+    nothing   → Left fakeErr -- ["DiemDbError::NotFound", "LedgerInfo for epoch", lsE epoch]
+    (just li) →
+      grd‖ li ^∙ liwsLedgerInfo ∙ liVersion /= version ≔
+           Left fakeErr --["epoch did not end at version", lsE epoch, lsVersion version]
+         ‖ is-nothing (li ^∙ liwsLedgerInfo ∙ liNextEpochState) ≔
+           Left fakeErr -- ["not an epoch change at version", lsVersion version]
+         ‖ otherwise≔
+           pure li
+--where
+-- here t = "LedgerStore":"getEpochEndingLedgerInfo":t
+
+getEpochState : LedgerStore → Epoch → Either ErrLog EpochState
+getEpochState self epoch = do
+  lcheck (epoch >? 0) (here' ("EpochState only queryable for epoch >= 1" {-∷ lsE epoch-} ∷ []))
+  case Map.lookup (epoch ∸ 1) (self ^∙ lsObmEpochToLIWS) of λ where
+    nothing → Left fakeErr --[ "DiemDbError::NotFound"
+                           --, "last LedgerInfo of epoch", lsE (epoch - 1) ]))
+    (just ledgerInfoWithSigs) →
+      maybeS (ledgerInfoWithSigs ^∙ liwsNextEpochState)
+             (Left fakeErr) --[ "last LedgerInfo in epoch must carry nextEpochState"
+                            --, lsE epoch, lsLI (ledgerInfoWithSigs^.liwsLedgerInfo) ]
+             pure
+ where
+  here' : List String → List String
+  here' t = "LedgerStore" ∷ "getEPochState" ∷ t
+
+-- iterator that yields epoch ending ledger infos
+-- starting from `start_epoch`
+-- ends at the one before `end_epoch`
+getEpochEndingLedgerInfoIter
+  : LedgerStore → Epoch → Epoch
+  → Either ErrLog EpochEndingLedgerInfoIter
+getEpochEndingLedgerInfoIter self startEpoch endEpoch =
+  EpochEndingLedgerInfoIter∙new <$> (foldM) go [] (fromToList startEpoch (endEpoch ∸ 1))
```

Comment:
The combination of the use of monus and a homegrown `fromToList` function without a clear description of its intended semantics risks misunderstanding or misuse of `fromToList` later.  These ranges would differ in Haskell and Agda if, say, `startEpoch` and `endEpoch` were negative and 0, respectively.  Conceivably, it could be correct and verified in Agda, but wrong in Haskell (e.g., if the Haskell code accidentally calculates a negative epoch).  Let's discuss the best way to address this.  Please at least include a comment about this before merging.  Thanks.

Reply:
Good point.  I added an edited version of your comment.